break up statements on acceptable 200 responses for resources

bobholt opened the pull request:
Break theses statements up for ease of understanding and testing.

Consider moving the `null` case to fetching relationships, as that seems to be the only place it could occur.

bobholt committed:
break up statements on acceptable 200 responses for resources

tkellen commented:
I'm :+1: to moving that null case into the fetch relationships section.

bobholt commented:
Note there is an existing statement [there](http://jsonapi.org/format/#fetching-relationships-responses-404) that could be replaced:

> If a relationship link URL exists but the relationship is empty, then `200 OK` **MUST** be returned, as described above.

tkellen commented:
double +1 to repurposing that.

bobholt committed:
move and expand on `200 OK` but `null`

bobholt commented:
Took a swing at it. Feel free to tweak/trash.

dgeb commented:
I feel like we lose clarity by moving the `null` case to the 404 section, especially since a 404 is not appropriate for the `null` case.

I'd rather turn the MUST in the 404 section into a non-normative note to eliminate the repetition of the same normative requirement.

tkellen commented:
Mind updating this so we can land @bobholt?

bobholt commented:
I guess I'm unsure how to proceed. I moved the `null` case to the relationship section, but left it in `200 OK`. I added the collection empty array case explicitly in the relationship section for symmetry. I left the `404` section alone.

I'll make the existing `404` statement non-normative, and we can take any further changes from there.

tkellen commented:
I think this looks good. @dgeb can you confirm? I think you might've been looking at a collapsed diff when you made your last comment.

dgeb commented on format/index.md:
```diff
@@ -809,6 +785,16 @@ value for [resource linkage], as described above for
+resources currently exist for that relationship, a server **MUST** respond to
+a successful fetch request with an empty array (`[]`) provided as primary data
+for the response document.
+
```
These two paragraphs are covered entirely by the paragraph above that reads:

>  The primary data in the response document **MUST** match the appropriate value for [resource linkage], as described above for [relationship objects][relationships].

dgeb commented on format/index.md:
```diff
@@ -887,8 +873,8 @@ a relationship link URL that does not exist.
-If a relationship link URL exists but the relationship is empty, then
-`200 OK` **MUST** be returned, as described above.
+> Note: If a relationship link URL exists but the relationship is empty, then
+`200 OK` will be returned, as described above.
```
I'm good with making this non-normative so that the normative statement is in one place :+1:

dgeb commented on format/index.md:
````diff
@@ -745,21 +736,6 @@ Content-Type: application/vnd.api+json
-  "data": null
-}
-```
-
````
I believe that the normative mention of the possible `null` response here is appropriate. It is not redundant with the "fetching relationships" section, since this section covers "fetching resources". These sections clarify that `null` is an appropriate response at a related resource URL (e.g. `/articles/1/author`).
